python/dgl/graphbolt/itemset.py

```diff
@@ -85,9 +85,10 @@ def __iter__(self) -> Iterator:
         if len(self._items) == 1:
             yield from self._items[0]
             return
-        zip_items = zip(*self._items)
-        for item in zip_items:
-            yield tuple(item)
+
+        items_len = self._items[0].shape[0]
```

Comment:
`shape` is not always available though it's `torch.Tensor` in most cases. we're assuming `item` is iterable. see the length check in `__len__()` in below lines. `len()` is more general, I think.

Reply:
And even `len()` is not always available. does `zip()` require to obtain the length? adding dependency on `len` will restrict the usage of `ItemSet`. the test case shown in https://github.com/dmlc/dgl/blob/50b05723ff1a379d10cd9b164a7cf54b32f4794c/tests/python/pytorch/graphbolt/test_itemset.py#L49 is still iterable though no length is available.

Reply:
how about adding an `if` statement here, if `__len__` is available, let's iterate over with explicit for loop, otherwise, use `zip()` instead.

Reply:
> `shape` is not always available though it's `torch.Tensor` in most cases. we're assuming `item` is iterable. see the length check in `__len__()` in below lines. `len()` is more general, I think.

`zip` doesn't require explicit length specification, but it does offer an optional `strict` parameter to control whether lengths should be pre-validated for a match.

The approach below is solid, making the `ItemSet` more robust.
> how about adding an if statement here, if `__len__` is available, let's iterate over with explicit for loop, otherwise, use `zip()` instead.

One question: What method should we use to check if `__len__` is available? Should we use `if isinstance(self._items[0], Sized):` or a `try` block?

Reply:
I think `if isinstance(self._items[0], Sized):` is ok.

Reply:
BTW, could you add `iterate logic` in https://github.com/dmlc/dgl/blob/50b05723ff1a379d10cd9b164a7cf54b32f4794c/tests/python/pytorch/graphbolt/test_itemset.py#L49. This will help us capture and assure the change in this PR. Or you could add another dedicated test case for this PR.

Reply:
Done